### PR TITLE
Make quantization modules work with DataParallel

### PR DESCRIPTION
**Pull request by vkuzo**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#37032 Make quantization modules work with DataParallel**

Summary:

DataParallel requires all params and buffers of child modules to be updated
in place because of how it implements model replication during the
forward pass (see https://github.com/pytorch/pytorch/pull/12671 for
context). Any params or buffers not updated in place are lost and not
propagated back to the master.

This diff updates all observers and fake_quant modules to do their parameter update in-place. This will enable static quant and QAT to work correctly with DataParallel.

Depends on https://github.com/pytorch/pytorch/pull/32684 and https://github.com/pytorch/pytorch/pull/37185.

Test Plan:

Script failed before and passes after the diff:
https://gist.github.com/vkuzo/78b06c01f23f98ee2aaaeb37e55f8d40

Added integration and unit tests to cover:

```
python test/test_quantization.py TestDistributed
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21206454](https://our.internmc.facebook.com/intern/diff/D21206454)

### CODE REVIEW

File: test/quantization/test_workflow_module.py [hunk 1]
```diff
@@ -670,3 +677,99 @@ def test_fq_serializable(self):
...
+            nn.Conv2d(1, 2, 3, stride=2, padding=1, bias=False),
+            nn.BatchNorm2d(2),
+            nn.AvgPool2d(14),
+            nn.Sigmoid(),
+            torch.quantization.DeQuantStub(),
+        )
+
+        torch.quantization.fuse_modules(model, [['1', '2', '3'], ['4', '5']], inplace=True)
+
+        model.qconfig = torch.quantization.get_default_qat_qconfig('fbgemm')
```
Comment (supriyar) [inline test/quantization/test_workflow_module.py:759]: Could you check to make sure fbgemm is in `supported_qengines` here?

File: test/quantization/test_workflow_module.py [hunk 2]
```diff
@@ -670,3 +677,99 @@ def test_fq_serializable(self):
...
         for key in state_dict:
             self.assertEqual(state_dict[key], loaded_dict[key])
+
+def _get_buffer_ids(module):
+    """
+    Object addresses stay constant if and only if all modifications are in-place
+    """
+    return [id(v) for k, v in module._buffers.items()]
+
+class TestDistributed(QuantizationTestCase):
```
Comment (supriyar) [inline test/quantization/test_workflow_module.py:687]: This is awesome, I think we should also add a test for DDP here at some point.

File: torch/quantization/fake_quantize.py [hunk 3]
```diff
@@ -79,7 +79,12 @@ def forward(self, X):
         if self.observer_enabled:
             self.activation_post_process(X.detach())
             _scale, _zero_point = self.calculate_qparams()
-            self.scale, self.zero_point = _scale.to(self.scale.device), _zero_point.to(self.zero_point.device)
+            _scale, _zero_point = _scale.to(self.scale.device), _zero_point.to(self.zero_point.device)
+            self.scale.resize_(_scale.shape)
```
Comment (supriyar) [inline torch/quantization/fake_quantize.py:83]: Why do we have to resize before calling `copy_`

Comment (vkuzo) [reply torch/quantization/fake_quantize.py:83]: this is because in the constructor the tensors are initialized to empty, without the knowledge of incoming shapes.  During the first iteration, the incoming shape has a different size from the current shape, so we have to resize to match because `copy_` requires equal size.  Before this diff, this was happening implicitly with direct assignment.

File: torch/quantization/observer.py [hunk 4]
```diff
@@ -284,6 +284,7 @@ def __init__(self, dtype=torch.quint8, qscheme=torch.per_tensor_affine,
     def forward(self, x_orig):
         r"""Records the running minimum and maximum of ``x``."""
         x = x_orig.detach()  # avoid keeping autograd tape
+        x = x.to(self.min_val.dtype)
```
Comment (raghuramank100) [inline torch/quantization/observer.py:287]: Why is this needed?

Comment (vkuzo) [reply torch/quantization/observer.py:287]: on lines 276/277, the min and max val buffers are initialized with the default floating point data type.  Since all updates are done in place, comparisons will crash unless the data types of both tensors being compared are equal.

File: torch/quantization/observer.py [hunk 5]
```diff
@@ -880,9 +894,13 @@ def forward(self, x_orig):
...
             min_val = torch.min(x)
             max_val = torch.max(x)
-            self.min_val = min_val
-            self.max_val = max_val
-            self.histogram = torch.histc(x, self.bins, min=min_val, max=max_val)
+            self.min_val.resize_(min_val.shape)
+            self.min_val.copy_(min_val)
+            self.max_val.resize_(max_val.shape)
+            self.max_val.copy_(max_val)
+            new_histogram = torch.histc(x, self.bins, min=min_val, max=max_val)
```
Comment (raghuramank100) [inline torch/quantization/observer.py:901]: Can use the version with an out argument instead:
torch.histc(input, bins=100, min=0, max=0, out=None)

File: torch/quantization/observer.py [hunk 6]
```diff
@@ -656,8 +668,10 @@ def forward(self, x_orig):
         else:
             min_vals = min_vals + self.averaging_constant * (torch.min(y, 1)[0] - min_vals)
             max_vals = max_vals + self.averaging_constant * (torch.max(y, 1)[0] - max_vals)
-        self.min_vals = min_vals
-        self.max_vals = max_vals
+        self.min_vals.resize_(min_vals.shape)
```
Comment (raghuramank100) [inline torch/quantization/observer.py:671]: Will using the out variant help with speed? For example, can we use this:
torch.add(input, other, out=None)

File: test/quantization/test_workflow_module.py (hunk 2 above)
Comment (raghuramank100) [reply test/quantization/test_workflow_module.py:687]: I agree fully. For the tests below, is there any value in making sure that they work with actual calls to dataparallel? The main value I can think of is if someone makes changes to dataparallel, we can catch that if our tests fail.

File: test/quantization/test_workflow_module.py [hunk 7]
```diff
@@ -670,3 +677,102 @@ def test_fq_serializable(self):
...
+
+    def test_observers_preserve_buffers(self):
+        """
+        Tests that observers only modify buffers in place. Note: this is important
+        because nn.DataParallel depends on this assumption to work correctly.
+        However, DataParallel does not expose IDs of the replicas, so we test it
+        without DataParallel in order to easily access the object IDs.
```
Comment (raghuramank100) [inline test/quantization/test_workflow_module.py:694]: Can we test with dataparallel?

Comment (vkuzo) [reply test/quantization/test_workflow_module.py:694]: The new test suite already has three tests:
* two unit tests which test the actual issue.  It's not possible to test this specific issue in DataParallel because the buffers of the replicas are hidden.
* an integration test, using DataParallel.

File: test/quantization/test_workflow_module.py (hunk 2 above)
Comment (vkuzo) [reply test/quantization/test_workflow_module.py:687]: I added comments in the code and commented regarding this on your comment below, hope that's ok.  I'm flexible.

File: torch/quantization/observer.py (hunk 6 above)
Comment (vkuzo) [reply torch/quantization/observer.py:671]: hmm, I took a look and it seems like only some flavors of `min/max` are supported with `=out`, so we wouldn't be able to apply that consistently throughout this diff.  Might make sense to keep it as is for simplicity, unless we are sure that this is a performance bottleneck. Thoughts/concerns?

**Review by raghuramank100 — APPROVED** [review]

File: torch/quantization/fake_quantize.py [hunk 8]
```diff
@@ -79,7 +79,12 @@ def forward(self, X):
...
+            self.scale.copy_(_scale)
```
Comment (fxmarty) [inline torch/quantization/fake_quantize.py:84]: Are these `copy_` operations dangerous in case the parameters are updated simultaneously over multiple GPUs (say two GPUs try to update "at the same time")?

Comment (fxmarty) [reply torch/quantization/fake_quantize.py:84]: Example showcasing dubious behavior

```python
import os
os.environ['CUDA_VISIBLE_DEVICES'] = '0, 1'

import torch
import torch.nn as nn

class M(nn.Module):
    def __init__(self):
        super(M, self).__init__()
        self.fc1 = nn.Linear(4, 3)
        self.fc2 = nn.Linear(3, 2)

    def forward(self, x):
        self.fc1.myattribute.copy_(self.net.fc1.myattribute + 3)
        print(self.fc1.myattribute)
        x = self.fc1(x)
        x = self.fc2(x)
        return x

globnet = M()

globnet.fc1.myattribute = torch.tensor(0)

input = torch.randn(4, 4, 4, 4)

device_ids = [0, 1]

data_scattered = nn.parallel.scatter(input, device_ids, dim=0)
# transfer the parameters to all GPUs
replicas = nn.parallel.replicate(globnet.cuda(), device_ids)

outputs = nn.parallel.parallel_apply(replicas, data_scattered)
gathered = nn.parallel.gather(outputs, 0)

print('end', globnet.net.fc1.myattribute)
```

This prints at times
```
tensor(3)
tensor(3)
end tensor(3)
```

and sometimes

```
tensor(6)
tensor(6)
end tensor(6)
```

What do you think?

Comment (fxmarty) [reply torch/quantization/fake_quantize.py:84]: ping @vkuzo
